### PR TITLE
darwin-builder: use port 31022 by default

### PR DESCRIPTION
**Pull request by Enzime**
###### Description of changes

- [darwin.builder: use port 31022 by default](https://github.com/Enzime/nixpkgs/commit/darwin-builder~3)
- [nixos/qemu-vm: use CA certificates from host](https://github.com/Enzime/nixpkgs/commit/darwin-builder~2)
- [darwin.builder: allow overriding configuration](https://github.com/Enzime/nixpkgs/commit/darwin-builder~1)
- [darwin.linux-builder: rename from darwin.builder](https://github.com/Enzime/nixpkgs/commit/darwin-builder)

This PR makes it so that `darwin.builder` no longer conflicts with `sshd` on macOS.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by roberth — CHANGES_REQUESTED** [review]
This has an unnecessary breaking change.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (Enzime) [reply nixos/modules/virtualisation/qemu-vm.nix]: Updated 👍

File: nixos/modules/profiles/macos-builder.nix
Comment (tjni) [inline nixos/modules/profiles/macos-builder.nix:239]: As I was reading this comment, I was asking myself what does this bootstrap problem entail and when might I run into it. If this comment is worth documenting, would it be possible to expand upon it?

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (tjni) [inline nixos/modules/virtualisation/qemu-vm.nix:892]: Could we accept an optional path to a certificate file instead? This way `qemu-vm` need not implicitly depend upon NIX_SSL_CERT_FILE.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (tjni) [inline nixos/modules/virtualisation/qemu-vm.nix:1035]: I'm a dummy when it comes to security, so this is a basic question: what's the implication on the guest VM if this is false? Will there be no certificates at all, and does that significantly limit the usability?

File: nixos/modules/profiles/macos-builder.nix
Comment (tjni) [reply nixos/modules/profiles/macos-builder.nix:70]: I also hope it's possible to separate properties like host port, memory, and others from the cached builder. I'm invested in this too so that I can use my Mac to test or review Linux packages and can spend some time on it too, but only if you desire it.

**Review by tjni — COMMENTED** [review]
Thank you for starting this effort and for your foresight on the benefit of a `nix-darwin` module. I just realized I wanted this today, and was glad to see some work on this already.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (tjni) [inline nixos/modules/virtualisation/qemu-vm.nix:176]: Is copying required (or better) than symlinking?

Comment (Enzime) [reply nixos/modules/virtualisation/qemu-vm.nix:176]: From my testing, I believe copying is required here as the symlink gets resolved within the VM not on the host side

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (Gabriella439) [reply nixos/modules/virtualisation/qemu-vm.nix:1035]: Instead of automatically disabling `security.pki` I think it should be treated as an error if the two options conflict.  In other words, something like:

```nix
{ assertions = [
    { assertion = !(config.services.pki.enable && config.virtualisation.useHostCerts);
      message = "…";
    }
  ];
}
```

File: nixos/modules/profiles/macos-builder.nix
Comment (Gabriella439) [reply nixos/modules/profiles/macos-builder.nix:70]: Yeah, changing the `hostPort` won't invalidate the cached VM build

I also feel like we should keep the default value to 22 if we can to avoid breaking existing configurations

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (roberth) [reply nixos/modules/virtualisation/qemu-vm.nix:1035]: I don't think PKI is something you can disable. The packages you install will rely on it anyway. They generally can't even be configured to work without it, so the idea that `security.pki.enable = false;` would disable this crucial security feature is not realistic, and not something anyone would want in real life systems. Probably not even in tests.

How about `security.pki.etc.enableCACert` or something like that?

I think it would also make more sense for the `ca.nix` module to be responsible for handling this use case. That way it can't be accidentally forgotten about when `ca.nix` is inevitably changed in a way that's not compatible with mounted files.

Comment (roberth) [reply nixos/modules/virtualisation/qemu-vm.nix:1035]: Besides mounted files, this module could support mutable ca files, which is similar but does not require virtualization or any kind of mount.
In both cases, the added value of this module is the knowledge of which files should be symlinked to the certificate file; whether that's a file that's composed from n immutable sources, or a single mutable or mounted file.

Comment (Enzime) [reply nixos/modules/virtualisation/qemu-vm.nix]: I've updated the name of the option to `security.pki.installCACerts`. I like the idea of supporting mutable CA files, however I would prefer not to do it in this PR as I think it would be out of the scope of this PR and potentially quite complex.

File: nixos/modules/profiles/macos-builder.nix
Comment (Enzime) [reply nixos/modules/profiles/macos-builder.nix]: Updated the comment :+1:

File: nixos/modules/profiles/macos-builder.nix
Comment (Enzime) [reply nixos/modules/profiles/macos-builder.nix:70]: I've updated this PR and added a way for consumers to be able to override `darwin.builder` more easily:

[darwin.builder: allow overriding configuration](https://github.com/NixOS/nixpkgs/commit/dac055259ad4b2972abbffd805b672e40d912c2a)

My preference would be changing the default so that future users don't have to deal with exposing the VM on port 22, and with the above change, existing users can choose to use the existing port or switch to a different port with the updated instructions in the documentation.

Comment (Enzime) [reply nixos/modules/profiles/macos-builder.nix:70]: Alternatively, we could not set a default port, and let the users configure this using an `override` or through the module system.

Although it would break the following command:

```
$ nix run nixpkgs#darwin.builder
```

Comment (roberth) [reply nixos/modules/profiles/macos-builder.nix:70]: We could rename it to `darwin.linuxBuilder` and inform users about the change using the deprecation message.

Not a repeatable solution though... Does anyone have other breaking changes in mind? Should we wait?

File: pkgs/top-level/darwin-packages.nix
Comment (roberth) [inline pkgs/top-level/darwin-packages.nix:244]:
```suggestion
      nixos.config.system.build.macos-builder-installer) { modules = [ ]; };
```

Consistent with `nixpkgsFlake.lib.nixosSystem` and `evalModules`. Use `++` in `imports`.

File: nixos/modules/profiles/macos-builder.nix
Comment (roberth) [reply nixos/modules/profiles/macos-builder.nix:70]: Oh actually we might change it again to use package name syntax instead of general identifier syntax: `linux-builder`. Awkward but useful? Maybe we should just give it a version at some point.

File: nixos/modules/virtualisation/qemu-vm.nix
Comment (roberth) [reply nixos/modules/virtualisation/qemu-vm.nix]: Yeah that's fine. I mentioned it as an example use case that might come later.

File: pkgs/top-level/darwin-packages.nix
Comment (roberth) [inline pkgs/top-level/darwin-packages.nix:224]: I think we should have a cross compiled version of the linux builder available. Gabriella's solution is great and robust / low maintenance because it reuses existing builds. However, as an alternate, potentially simpler solution to the bootstrapping problem, it would be great to have a cross compiled configuration on Hydra and in the cache (although the cache is not a requirement in this case, which is actually the whole point, it's always nice to have some cache).


```suggestion
  linuxBuilder_cross = linuxBuilder.override (o: { modules = o.modules ++ [{
    nixpkgs.buildPlatform = stdenv.buildPlatform;
  }]; });
```

I think you'd also have to change `nixpkgs.system` in the default config to `nixpkgs.hostPlatform`.

(These `*Platform` ones are new options that are a bit friendlier than the old `crossSystem`/`localSystem` combination.)

File: pkgs/top-level/darwin-packages.nix
Comment (Enzime) [reply pkgs/top-level/darwin-packages.nix]: Updated :+1:

File: nixos/modules/profiles/macos-builder.nix
Comment (Enzime) [reply nixos/modules/profiles/macos-builder.nix:70]: I'm happy to rename this and add the deprecation message, is there a preference between `linuxBuilder` or `linux-builder`?

File: pkgs/top-level/darwin-packages.nix
Comment (Enzime) [reply pkgs/top-level/darwin-packages.nix:224]: I'm not too familiar with cross building, happy to investigate in another PR if that's okay with you.

Comment (roberth) [reply pkgs/top-level/darwin-packages.nix:224]: Of course; just wanted to put the idea out there.

File: nixos/modules/profiles/macos-builder.nix
Comment (roberth) [reply nixos/modules/profiles/macos-builder.nix:70]: Let's go with `linux-builder`.

File: pkgs/top-level/aliases.nix
Comment (Enzime) [inline pkgs/top-level/aliases.nix:354]: @roberth any suggestions on how to write this alias? This fails with:

```
error:
       … while evaluating the attribute 'darwin.overrideScope'

         at /nix/store/2l7lgxsidza81llpq5imhxf3a7cdk1zl-source/pkgs/stdenv/darwin/default.nix:651:9:

          650|
          651|         darwin = super.darwin.overrideScope (_: _: {
             |         ^
          652|           inherit (darwin) dyld ICU Libsystem Csu libiconv rewrite-tbd;

       … while evaluating a branch condition

         at /nix/store/2l7lgxsidza81llpq5imhxf3a7cdk1zl-source/pkgs/top-level/aliases.nix:33:5:

           32|   removeDistribute = alias: with lib;
           33|     if isDerivation alias then
             |     ^
           34|       dontDistribute alias

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Alias darwin is still in all-packages.nix
```

Comment (roberth) [reply pkgs/top-level/aliases.nix:354]: I think `aliases.nix` is only for top level aliases, but `builder` is inside the `darwin` attribute, and `darwin` is by itself not an alias.

You could still support `config.allowAliases` by manually making your attribute optional based on that bool.

Comment (Enzime) [reply pkgs/top-level/aliases.nix]: Updated :+1:

File: doc/builders/special/darwin-builder.section.md
Comment (roberth) [inline doc/builders/special/darwin-builder.section.md:7]:
```suggestion
The builder runs on host port 31022 by default.
You can change it by overriding `virtualisation.darwin-builder.hostPort`.
See the [example](#ex-darwin-linux-builder-flake).
```

We try to do one sentence per line. A single newline is not rendered as a line break.

Also down below (can't make a proper suggestion because it's not touched)

```diff
-## Example flake usage
+## Example flake usage {#ex-darwin-linux-builder-flake}
```

File: pkgs/top-level/darwin-packages.nix
Comment (roberth) [inline pkgs/top-level/darwin-packages.nix:247]: They wouldn't know what to look for, but we can make them aware of the change here and still refer to the instructions for details.

```suggestion
  builder = throw "'darwin.builder' has been changed and renamed to 'darwin.linux-builder'. The default ssh port is now 31022. Please update your configuration or override the port back to 22. See https://nixos.org/manual/nixpkgs/unstable/#sec-darwin-builder"; # added 2023-07-04
```

**Review by roberth — COMMENTED** [review]
Just need to complete the docs, otherwise lgtm.

File: doc/builders/special/darwin-builder.section.md
Comment (Enzime) [reply doc/builders/special/darwin-builder.section.md]: Updated :+1:

File: pkgs/top-level/darwin-packages.nix
Comment (Enzime) [reply pkgs/top-level/darwin-packages.nix]: Updated :+1:

**Review by roberth — APPROVED** [review]
LGTM if someone could test this.

**Review by emilytrau — APPROVED** [review]
Tested to work on `aarch64-darwin` 🎉

**Pull request merged**
